File: Array/2195.Append_K_Integers_With_Minimal_Sum/Python/2195.Append_K_Integers_With_Minimal_Sum.py
class Solution(object):
    def minimalKSum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        #===============================#
        # Greedy-based traversal method #
        #===============================#

        ############
        #Initialize
        ##### Length of nums array #####
        len_nums = len(nums)

        ##### Record dictionary/hashmap #####
        record_dict = {}

        ##### Result minimun summary value #####
        res_min_sum = ((k * (k + 1)) // 2)


        ##############################################################
        #Greedy-based loop traversal with recorded dictionary/hashmap
        ##### Step 1: Record indexed-value with dictionary/hashmap #####
        for nums_idx in range(len_nums):
            record_dict[nums[nums_idx]] = True #Keep updating/recording
        
        ##### Step 2: Looped-traversal with recorded dictionary/hashmap #####
        for num in sorted(record_dict.keys()):

            ##### Check if the current indexed-value is valid or not #####
            if (num > k):
                break

            res_min_sum -= num #Keep pdating/reducing

            res_min_sum += (k + 1) #Keep updating/accumulating

            k += 1 #Keep updating/accumulating

        return res_min_sum

File: Array/2195.Append_K_Integers_With_Minimal_Sum/C++/2195.Append_K_Integers_With_Minimal_Sum.cpp
class Solution {
public:
    long long minimalKSum(vector<int> & nums, int k) 
    {
        //==============================================//
        // Input type:                                  //
        //   - vector<int> ref. (nums vector)           //
        //   - int (k-th ranges)                        //
        // Return type:                                 //
        //   - long long (result minimun summary value) //
        //==============================================//

        /*Initialize*/
        ///// Length of nums vector /////
        int len_nums = nums.size();

        ///// Record dictionary, hashmap /////
        map<int, bool> record_map;

        ///// Result minimun summary value /////
        long long res_min_sum = (((long long)(k) * ((long long)(k) + 1)) / 2);


        /*Greedy-based loop traversal with recorded dictionary, hashmap*/
        ///// Step 1: Record indexed-value with dictionary, hashmap /////
        for (int nums_idx = 0; (nums_idx < len_nums); (nums_idx++))
        {
            record_map[nums[nums_idx]] = true;
        }

        ///// Step 2: Looped-traversal with recorded dictionary, hashmap /////
        for (auto record_ele = record_map.begin(); (record_ele != record_map.end()); (record_ele++)) //Whole
        {
            ///// Check if the current indexed-value matched conditions or not /////
            if ((record_ele->first) <= k)
            {
                res_min_sum -= (record_ele->first); //Keep updating, reducing

                res_min_sum += (++k); //Keep updating, accumulating
            }
            else { break; }

        } //Whole

        return res_min_sum;
    }
};
